packages/svelte/src/compiler/phases/2-analyze/css/css-prune.js

```diff
@@ -1,261 +1,387 @@
-const UNKNOWN_SELECTOR = 'UNKNOWN_SELECTOR';
-
-const NodeExist = /** @type {const} */ ({
-	Probably: 0,
-	Definitely: 1
-});
+const NODE_PROBABLY_EXISTS = 0;
+const NODE_DEFINITELY_EXISTS = 1;
 
 const whitelist_attribute_selector = new Map([
 	['details', ['open']],
 	['dialog', ['open']]
 ]);
 
+/** @type {import('#compiler').Css.Combinator} */
+const descendant_combinator = {
+	type: 'Combinator',
+	name: ' ',
+	start: -1,
+	end: -1
+};
+
+/** @type {import('#compiler').Css.RelativeSelector} */
+const nesting_selector = {
+	type: 'RelativeSelector',
+	start: -1,
+	end: -1,
+	combinator: null,
+	selectors: [
+		{
+			type: 'NestingSelector',
+			name: '&',
+			start: -1,
+			end: -1
+		}
+	],
+	metadata: {
+		is_global: false,
+		is_host: false,
+		is_root: false,
+		scoped: false
+	}
+};
+
 /**
  *
  * @param {import('#compiler').Css.StyleSheet} stylesheet
  * @param {import('#compiler').RegularElement | import('#compiler').SvelteElement} element
  */
 export function prune(stylesheet, element) {
-	/** @type {State} */
-	const state = { stylesheet, element };
-
-	walk(stylesheet, state, visitors);
+	walk(stylesheet, { stylesheet, element }, visitors);
 }
 
 /** @type {import('zimmerframe').Visitors<import('#compiler').Css.Node, State>} */
 const visitors = {
 	ComplexSelector(node, context) {
-		context.next();
+		const selectors = truncate(node);
+		const inner = selectors[selectors.length - 1];
 
-		const i = node.children.findLastIndex((child) => {
-			return !child.metadata.is_global && !child.metadata.is_host && !child.metadata.is_root;
-		});
+		if (node.metadata.rule?.metadata.parent_rule) {
+			const has_explicit_nesting_selector = selectors.some((selector) =>
+				selector.selectors.some((s) => s.type === 'NestingSelector')
+			);
 
-		const relative_selectors = node.children.slice(0, i + 1);
+			if (!has_explicit_nesting_selector) {
+				selectors[0] = {
+					...selectors[0],
+					combinator: descendant_combinator
+				};
 
-		if (apply_selector(relative_selectors, context.state.element, context.state.stylesheet)) {
+				selectors.unshift(nesting_selector);
+			}
+		}
+
+		if (
+			apply_selector(
+				selectors,
+				/** @type {import('#compiler').Css.Rule} */ (node.metadata.rule),
+				context.state.element,
+				context.state.stylesheet
+			)
+		) {
+			mark(inner, context.state.element);
 			node.metadata.used = true;
 		}
-	},
-	RelativeSelector(node, context) {
-		// for now, don't visit children (i.e. inside `:foo(...)`)
-		// this will likely change when we implement `:is(...)` etc
+
+		// note: we don't call context.next() here, we only recurse into
+		// selectors that don't belong to rules (i.e. inside `:is(...)` etc)
+		// when we encounter them below
 	}
 };
 
+/**
+ * Discard trailing `:global(...)` selectors, these are unused for scoping purposes
+ * @param {import('#compiler').Css.ComplexSelector} node
+ */
+function truncate(node) {
+	const i = node.children.findLastIndex(({ metadata }) => {
```

Comment:
Is it possible that `i` might be `-1` here?

Reply:
Yes, and that's fine because we do `array.slice(0, -1 + 1)` which gives us an empty array, which is exactly what we want